src/testing/h_refinement_study_isentropic_vortex.cpp

```diff
@@ -0,0 +1,249 @@
+{}
+
+
+
+template <int dim, int nstate>
+Parameters::AllParameters HRefinementStudyIsentropicVortex<dim,nstate>::reinit_params_and_refine(int refinement) const
+{
+     PHiLiP::Parameters::AllParameters parameters = *(this->all_parameters);
+     
+     parameters.flow_solver_param.number_of_grid_elements_per_dimension *= pow(2, refinement);
+     parameters.flow_solver_param.unsteady_data_table_filename += std::to_string(parameters.flow_solver_param.number_of_grid_elements_per_dimension);
+     return parameters;
+}
+
+template <int dim, int nstate>
+void HRefinementStudyIsentropicVortex<dim,nstate>::calculate_Lp_error_at_final_time_wrt_function(double &Lp_error_density, 
+        double &Lp_error_pressure,
+        std::shared_ptr<DGBase<dim,double>> dg,
+        const Parameters::AllParameters parameters,
+        double final_time,
+        int norm_p) const
+{
+    //generate exact solution at final time
+    std::shared_ptr<ExactSolutionFunction<dim,nstate,double>> exact_solution_function;
+    exact_solution_function = ExactSolutionFactory<dim,nstate,double>::create_ExactSolutionFunction(parameters.flow_solver_param, final_time);
+    int overintegrate = 10;
+
+    // For Euler, compare only density or pressure
+    // deal.ii compute_global_error() does not interface simply
+    // Therefore, overintegrated error calculation is coded here
+    
+    // Need to do MPI sum manaully, so declaring a new local error
+    double Lp_error_density_local = 0;
+    double Lp_error_pressure_local = 0;
+    
+    std::shared_ptr< Physics::Euler<dim,dim+2,double> > euler_physics = std::dynamic_pointer_cast<Physics::Euler<dim,dim+2,double>>(
+            Physics::PhysicsFactory<dim,dim+2,double>::create_Physics(&parameters));
+    dealii::QGauss<dim> quad_extra(dg->max_degree+1+overintegrate);
+    dealii::FEValues<dim,dim> fe_values_extra(*(dg->high_order_grid->mapping_fe_field), dg->fe_collection[dg->max_degree], quad_extra,
+                                              dealii::update_values | dealii::update_gradients | dealii::update_JxW_values | dealii::update_quadrature_points);
+
+    const unsigned int n_quad_pts = fe_values_extra.n_quadrature_points;
+    std::array<double,nstate> soln_at_q;
+
+    std::vector<dealii::types::global_dof_index> dofs_indices (fe_values_extra.dofs_per_cell);
+    for (auto cell : dg->dof_handler.active_cell_iterators()) {
+        if (!cell->is_locally_owned()) continue;
+        fe_values_extra.reinit (cell);
+        cell->get_dof_indices (dofs_indices);
+
+        for (unsigned int iquad=0; iquad<n_quad_pts; ++iquad) {
+
+            std::fill(soln_at_q.begin(), soln_at_q.end(), 0.0);
+
+            for (unsigned int idof=0; idof<fe_values_extra.dofs_per_cell; ++idof) {
+                const unsigned int istate = fe_values_extra.get_fe().system_to_component_index(idof).first;
+                soln_at_q[istate] += dg->solution[dofs_indices[idof]] * fe_values_extra.shape_value_component(idof, iquad, istate);
```

Comment:
If max_degree>actual_degree, `dofs_indices[idof]` can give different dofs than that of the current cell, which might give incorrect soln_at_q (or the dof might not even be locally owned). Not sure if it will work when dg->max_degree is set to a higher value. If time permits, try running it in debug to see if it triggers any Assert statement in dealii.

Reply:
This is good to keep in mind for future development. Personally, I don't foresee myself using this test for non-uniform p. Therefore, checking for uniform p should be sufficient here.

I have left a comment in the abort statement that there may be issues. Thanks for pointing this out.

Reply:
Yes, there shouldn't be any issues if `flow_solver_param.max_poly_degree_for_adaptation` isn't used in parameters file.